src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,188 @@
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: i64 = 3;
+
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct CheckAboutToExpireToken;
+
+impl BackgroundJob for CheckAboutToExpireToken {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let mut conn = env.connection_pool.get()?;
+        // Check if the token is about to expire
+        // If the token is about to expire, trigger a notification.
+        check(&env.emails, &mut conn)
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens =
+        crate::models::token::ApiToken::find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    // Batch send notifications in transactions.
+    const BATCH_SIZE: usize = 100;
+    for chunk in expired_tokens.chunks(BATCH_SIZE) {
+        conn.transaction(|conn| {
+            for token in chunk {
+                // Send notification.
+                let user = User::find(conn, token.user_id)?;
+                let Some(recipient) = user.email(conn)? else {
+                    return Err(anyhow!("No address found"));
+                };
+                let email = ExpiryNotificationEmail {
+                    name: &user.gh_login,
+                    token_name: &token.name,
+                };
+                match emails.send(&recipient, email) {
+                    Ok(_) => {
+                        // Update the token to prevent duplicate notifications.
+                        diesel::update(token)
+                            .set(
+                                crate::schema::api_tokens::expiry_notification_at
+                                    .eq(now.nullable()),
+                            )
+                            .execute(conn)?;
+                    }
+                    Err(e) => {
+                        error!("Failed to send email: {:?} to {}", e, recipient);
+                    }
+                }
+            }
+            Ok::<_, anyhow::Error>(())
+        })?;
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+struct ExpiryNotificationEmail<'a> {
+    name: &'a str,
+    token_name: &'a str,
+}
+
+impl<'a> Email for ExpiryNotificationEmail<'a> {
+    const SUBJECT: &'static str = "Your token is about to expire";
+
+    fn body(&self) -> String {
+        format!(
+            r#"Hi @{},
+
+We noticed your token "{}" will expire in about {} days.
```

Comment:
we might want to serialize the exact date instead of "in about 3 days". maybe something like:

> Your "foo" crates.io API token will expiry on 2024-04-15.

Reply:
If we use the date directly, should we consider the time zone issue? Or do we just say it is UTC time?

Reply:
good point. I guess since we're a developer-focused service it's probably fine to use UTC and the full ISO8601 format (2024-04-15T12:34:56Z).

Reply:
Or even be explicit with the offset `+00:00`?

Reply:
Sounds good to me. I will update it.